src/Mobs/Monster.cpp

```diff
@@ -435,19 +412,54 @@ void cMonster::CalcLeashActions()
 			SetLeashToPos(nullptr);
 		}
 	}
-	else if (IsLeashed())  // Mob is already leashed to an entity: follow it.
+
+	if (!IsLeashed())
+	{
+		return;
+	}
+
+	static const double CloseFollowDistance = 1.8;   // The closest the mob will path towards the leashed to entity
+	static const double LeashNaturalLength  = 5.0;   // The closest the mob is actively pulled towards the entity
+	static const double LeashMaximumLength  = 10.0;  // Length where the leash breaks
+	static const double LeashSpringConstant = 20.0;  // How stiff the leash is
+
+	const auto LeashedToPos = m_LeashedTo->GetPosition();
+	const auto Displacement = LeashedToPos - GetPosition();
+	const auto Distance = Displacement.Length();
+	const auto Direction = Displacement.NormalizeCopy();
+
+	// If the leash is over-extended, break the leash:
+	if (Distance > LeashMaximumLength)
+	{
+		LOGD("Leash broken (distance)");
+		Unleash(false);
+		return;
+	}
+
+	// If the mob isn't following close enough, pull the mob towards the leashed to entity:
+	if (Distance > LeashNaturalLength)
 	{
-		// TODO: leashed mobs in vanilla can move around up to 5 blocks distance from leash origin
-		MoveToPosition(m_LeashedTo->GetPosition());
+		// Accelerate monster towards the leashed to entity:
+		const auto Extension = Distance - LeashNaturalLength;
+		auto Acceleration = Direction * (Extension * LeashSpringConstant);
 
-		// If distance to target > 10 break leash
-		Vector3f a_Distance(m_LeashedTo->GetPosition() - GetPosition());
-		double Distance(a_Distance.Length());
-		if (Distance > 10.0)
+		// Stop mobs from floating up when on the ground
+		if (IsOnGround() && (Acceleration.y < std::abs(GetGravity())))
 		{
-			LOGD("Leash broken (distance)");
-			Unleash(false);
+			Acceleration.y = 0.0;
 		}
+
+		// Apply the acceleration
+		using namespace std::chrono;
+		AddSpeed(Acceleration * duration_cast<duration<double>>(a_Dt).count());
+	}
+
+	// Passively follow the leashed to entity:
+	if (Distance > CloseFollowDistance)
+	{
+		const Vector3d TargetBlock((LeashedToPos - Direction * CloseFollowDistance).Floor());
+		// Move to centre of target block face
+		MoveToPosition(TargetBlock + Vector3d{ 0.5, 0.0, 0.5 });
```

Comment:
I'm not too familiar with the entity movement calculations, but is this movement cancelled when the leashed entity gets within the CloseFollowDistance? I feel like this could end up with mobs sitting on top of the player again.

Reply:
You're right that the movement isn't cancelled because `MoveToPosition` isn't called.  However the mobs aren't pathing towards the block the player is standing on in the first place. Before rounding, the target block is the nearest point on a sphere of radius `CloseFollowDistance` centred on the the leashed to entity.

If you're wondering why the if is necessary then: it actually stops leashed mobs from walking away when the player walks within the `CloseFollowDistance`

Reply:
Thanks for explaining, I understand it now and the logic looks good.